File: Part1_강의자료/Part1_연습문제_풀이/1300_K번째수/1300.py
N = int(input())
K = int(input())

l = 1
r = N * N
ans = -1
while l <= r:
    m = (l + r) // 2
    count = 0
    for i in range(1, N+1):
        count += min(m // i, N)
    if count >= K:
        ans = m
        r = m - 1
    else:
        l = m + 1
print(ans)

File: Part1_강의자료/Part1_연습문제_풀이/1300_K번째수/1300.cpp
#include <stdio.h>
#include <algorithm>
using namespace std;
typedef long long ll;

int main()
{
    int N, K;

    scanf("%d", &N);
    scanf("%d", &K);

    ll l = 1;
    ll r = (ll)N * N;
    ll ans = -1;
    while (l <= r)
    {
        ll m = (l + r) / 2;
        ll count = 0;
        for (int i = 1; i <= N; i++)
            count += min(m / i, (ll)N);
        if (count >= K)
        {
            ans = m;
            r = m - 1;
        }
        else
            l = m + 1;
    }
    printf("%lld\n", ans);
    return 0;
}
